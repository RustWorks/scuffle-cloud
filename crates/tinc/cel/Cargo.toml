[package]
name = "tinc-cel"
version = "0.1.0"
edition = "2024"
description = "A CEL Type used by tinc-build and tinc for validating protobuf messages"
repository = "https://github.com/scufflecloud/scuffle"
authors = ["Scuffle <opensource@scuffle.cloud>"]
readme = "README.md"
documentation = "https://docs.rs/tinc-build"
license = "MIT OR Apache-2.0"
keywords = ["grpc", "protobuf", "tonic", "codegen"]

[dependencies]
num-traits = "0.2.19"
bytes = "1.5"
float-cmp = "0.10"
thiserror = "2"
chrono = "0.4"
regex = "1"
linkme = "0.3"
fmtools = "0.1"
uuid = "1"
url = "2"
email_address = "0.2.9"
scuffle-workspace-hack.workspace = true

[features]
runtime = []

