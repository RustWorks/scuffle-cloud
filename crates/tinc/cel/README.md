<!-- cargo-sync-rdme title [[ -->
# tinc-cel
<!-- cargo-sync-rdme ]] -->

> [!WARNING]  
> This crate is under active development and may not be stable.

<!-- cargo-sync-rdme badge [[ -->
![License: MIT OR Apache-2.0](https://img.shields.io/crates/l/tinc-cel.svg?style=flat-square)
[![docs.rs](https://img.shields.io/docsrs/tinc-cel.svg?logo=docs.rs&style=flat-square)](https://docs.rs/tinc-cel)
[![crates.io](https://img.shields.io/crates/v/tinc-cel.svg?logo=rust&style=flat-square)](https://crates.io/crates/tinc-cel)
[![GitHub Actions: ci](https://img.shields.io/github/actions/workflow/status/scufflecloud/scuffle/ci.yaml.svg?label=ci&logo=github&style=flat-square)](https://github.com/scufflecloud/scuffle/actions/workflows/ci.yaml)
[![Codecov](https://img.shields.io/codecov/c/github/scufflecloud/scuffle.svg?label=codecov&logo=codecov&style=flat-square)](https://codecov.io/gh/scufflecloud/scuffle)
<!-- cargo-sync-rdme ]] -->

---

<!-- cargo-sync-rdme rustdoc [[ -->
Currently this is a fully private api used by `tinc` and `tinc-build` to
compile and execute [CEL](https://cel.dev/) expressions.
Check out the [changelog](./CHANGELOG.md).

### Feature flags

* **`runtime`** —  Enables runtime enum type indexing
* **`docs`** —  Enables changelog and documentation of feature flags

### License

This project is licensed under the MIT or Apache-2.0 license.
You can choose between one of them if you use this work.

`SPDX-License-Identifier: MIT OR Apache-2.0`
<!-- cargo-sync-rdme ]] -->
