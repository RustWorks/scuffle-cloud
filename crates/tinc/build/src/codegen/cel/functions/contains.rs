use cel_interpreter::{ExecutionError, FunctionContext};
use quote::quote;
use syn::parse_quote;

use super::Function;
use crate::codegen::cel::compiler::{CompileError, CompiledExpr, CompilerCtx};
use crate::codegen::cel::types::CelType;
use crate::types::{ProtoModifiedValueType, ProtoType, ProtoValueType};

#[derive(Debug, Clone, Default)]
pub struct Contains;

// this.contains(arg) -> arg in this
impl Function for Contains {
    fn name(&self) -> &'static str {
        "contains"
    }

    fn compile(&self, ctx: CompilerCtx) -> Result<CompiledExpr, CompileError> {
        let Some(this) = &ctx.this else {
            return Err(CompileError::MissingTarget {
                func: self.name(),
                message: "this is required when calling the contains function".to_string(),
            });
        };

        if ctx.args.len() != 1 {
            return Err(CompileError::InvalidFunctionArgumentCount {
                func: self.name(),
                expected: 1,
                got: ctx.args.len(),
            });
        }

        let arg = ctx.resolve(&ctx.args[0])?.to_cel()?;

        if let CelType::Proto(ProtoType::Modified(
            ProtoModifiedValueType::Repeated(item) | ProtoModifiedValueType::Map(item, _),
        )) = &this.ty
        {
            if !matches!(item, ProtoValueType::Message { .. } | ProtoValueType::Enum(_)) {
                let op = match &this.ty {
                    CelType::Proto(ProtoType::Modified(ProtoModifiedValueType::Repeated(_))) => {
                        quote! { array_contains }
                    }
                    CelType::Proto(ProtoType::Modified(ProtoModifiedValueType::Map(_, _))) => {
                        quote! { map_contains }
                    }
                    _ => unreachable!(),
                };

                return Ok(CompiledExpr {
                    expr: parse_quote! {
                        ::tinc::__private::cel::#op(
                            #this,
                            #arg,
                        )
                    },
                    ty: CelType::Proto(ProtoType::Value(ProtoValueType::Bool)),
                });
            }
        }

        let this = this.clone().to_cel()?;

        Ok(CompiledExpr {
            expr: parse_quote! {
                ::tinc::__private::cel::CelValue::cel_contains(
                    #this,
                    #arg,
                )?
            },
            ty: CelType::Proto(ProtoType::Value(ProtoValueType::Bool)),
        })
    }

    fn interpret(&self, fctx: &FunctionContext) -> Result<cel_interpreter::Value, ExecutionError> {
        let Some(this) = &fctx.this else {
            return Err(ExecutionError::missing_argument_or_target());
        };

        if fctx.args.len() != 1 {
            return Err(ExecutionError::invalid_argument_count(1, fctx.args.len()));
        }

        let arg = fctx.ptx.resolve(&fctx.args[0])?;

        match (this, arg) {
            (cel_interpreter::Value::String(s), cel_interpreter::Value::String(t)) => {
                Ok(cel_interpreter::Value::Bool(s.contains(t.as_str())))
            }
            (cel_interpreter::Value::Bytes(s), cel_interpreter::Value::Bytes(t)) => {
                Ok(cel_interpreter::Value::Bool(s.windows(t.len()).any(|w| w == t.as_slice())))
            }
            (cel_interpreter::Value::List(s), value) => Ok(cel_interpreter::Value::Bool(s.contains(&value))),
            (cel_interpreter::Value::Map(s), value) => {
                let key: Option<cel_interpreter::objects::Key> = value.clone().try_into().ok();
                Ok(cel_interpreter::Value::Bool(key.is_some_and(|k| s.get(&k).is_some())))
            }
            _ => Ok(cel_interpreter::Value::Bool(false)),
        }
    }
}
