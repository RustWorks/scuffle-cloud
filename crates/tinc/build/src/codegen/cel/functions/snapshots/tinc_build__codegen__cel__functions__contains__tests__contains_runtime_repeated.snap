---
source: crates/tinc/build/src/codegen/cel/functions/contains.rs
expression: result
---
Runtime(
    RuntimeCompiledExpr {
        ty: Proto(
            Value(
                Bool,
            ),
        ),
        expr: ::tinc::__private::cel::array_contains(
            input,
            ::tinc::__private::cel::CelValue::String(
                ::tinc::__private::cel::CelString::Borrowed("value"),
            ),
        ),
    },
)
